tbtc-ts/src/index.ts

```diff
@@ -143,6 +145,23 @@ export interface TBTC {
     depositData: DepositData[],
     mainUtxo?: UnspentTransactionOutput
   ): Promise<void>
+
+  /**
+   * Constructs data with the proof that a given sweep transaction is included
+   * in the Bitcoin blockchain and has the required number of confirmations as
+   * well as information on the transaction that is needed for processing
+   * on-chain.
+   * @param txId - ID of the sweep transaction.
+   * @param confirmations - Required number of confirmations for the transaction.
+   * @param bitcoinClient - Bitcoin client used to interact with the network.
+   * @returns Data on a sweep transaction with the inclusion proof and
+   *          information needed for processing on-chain.
+   */
+  constructSweepProof(
```

Comment:
I think we should not tie getting transaction proof with sweeping. Constructing proofs of BTC transactions is generic and can be used for redemption or moving funds proofs as well. 

**My proposition**

1. Remove `constructSweepProof` from `TBTC` and expose `proveSweep` function with the following signature:
```
proveSweep(transactionHash: string, bitcoinClient: BitcoinClient): Promise<void>
```
Note we don't need `confirmations` (more details below) and I think we should use `transactionHash` not  `txId` to be consistent with the way we define a `Transaction` in `bitcoin.ts`.

2. Rename `sweepProof.ts` to just `proof.ts`. Make all the code generic by removing all `sweep` prefixes and suffixes. The `proof.ts` file should expose just one function:
```
createTransactionProof(
    transactionHash: string, 
    confirmations: number, 
    bitcoinClient: BitcoinClient
): Promise<Transaction & Proof>
```
The `Proof` interface should be the same as `TxProof` in `bitcoin.ts`. I'd cut the `Tx` prefix since we are in Bitcoin context so `Proof` should be clear enough. Regarding `TransactionMerkleBranch` I think it should remain as is because it seems to be at the right place.

3. Introduce `bridge.ts` file where `Bridge` interface will live. For now, the `Bridge` interface should expose one method `submitSweepProof` with same signature as the function in `Bridge.sol` contract and probably a getter of `txProofDifficultyFactor` constant (since it's immutable we can cache it somehow). We don't need real implementation for now (apart from a test mock). Having the contract binding hidden behind an interface will give us more flexibility in the future and will prevent leaking contract details into `tbtc-ts` library.

4. Place implementation of `proveSweep` right into `sweep.ts` file. The actual implementation should call `txProofDifficultyFactor` to obtain the required confirmations factor, call `createTransactionProof` from `proof.ts` to build the proof and submit the proof using `submitSweepProof` from `bridge.ts`

**Notice about bitcoin.ts interfaces**

Regarding `bitcoin.ts` interfaces, I think we should be very careful with adding additional fields to Bitcoin types. They should represent actual Bitcoin types (https://developer.bitcoin.org/reference/transactions.html). That said, I think we should just focus on `Transaction` interface, i.e. add `version` and `locktime` fields as numbers but not hex strings. So, at the moment, we have two representations of the Bitcoin transaction:
- `Transaction` is a deserialized form and can be used by CLIs, dApps, etc. easily
- `RawTransaction` is a serialized form and is used to directly interact with the Bitcoin chain. Bear in mind there are two formats of serialization (https://hongchao.me/anatomy-of-raw-bitcoin-transaction) and that has important consequences.

We still need a representation to interact with Ethereum which contains the four fields (version, inputs, outputs, locktime) serialized and distinguishable. We can't just throw them to `RawTransaction` because of two possible serialization formats where the four fields can represent only the basic one but not the extended format. However, we can always obtain the four fields by decomposing the `RawTransaction`. Having that said, I think we should design the following chain:
```
Transaction <--CONVERT--> RawTransaction --DECOMPOSE--> {version, inputs, outputs, locktime}
```
In the future, we can write `TransactionConverter` class that will handle conversions between `Transaction` and `RawTransaction`. For now, we need a `RawTransactionDecomposer` class or `decomposeRawTransaction` function that will handle the decomposition and transform a `RawTransaction` into the four fields needed on the Ethereum side.

What do you think?

Reply:
Done.